src/memory_region/local.rs

```diff
@@ -21,7 +27,24 @@ pub trait LocalMrReadAccess: MrAccess {
 
     /// Get the local key
     fn lkey(&self) -> u32;
+
+    /// New a token with specified timeout
+    #[inline]
+    fn token_with_timeout(&self, timeout: Duration) -> Option<MrToken> {
+        SystemTime::now().checked_add(timeout).map_or_else(
```

Comment:
cool, I wil have a try.

Reply:
I tried some tests and found that `rdtsc` is faster if we just use `now()` and `elapsed()` to measure time in one machine.

But in our scene, we need to transfer ddl between multiple machines, so we have to do some conversions and additions 
with `Duration`.

`SystemTime` or `clock_gettime (Realtime Clock)` is faster in our scene if my test is correct.

Output:
  minstant: 45 ns
  systemtime: 31 ns

```
    fn main() {
        let timeout: Duration = Duration::new(1, 1);
        println!("minstant: {:?} ns", minstant(timeout));
        println!("systemtime: {:?} ns", systime(timeout));
    }

    fn minstant(timeout: Duration) -> u128 {
        let anchor = Anchor::new();
        let now = Instant::now();
        for _ in 0..1_000_000 {
            let _ddl = Instant::now()
                .as_unix_nanos(&anchor)
                .checked_add(timeout.as_nanos().cast())
                .unwrap();
        }
        now.elapsed().as_nanos() / 1_000_000
    }

    fn systime(timeout: Duration) -> u128 {
        let now = Instant::now();
        for _ in 0..1_000_000 {
            let _ddl = SystemTime::now().checked_add(timeout).unwrap();
        }
        now.elapsed().as_nanos() / 1_000_000
    }
```

Reply:
I add a commit for the try and some other changes mentioned in comments of this pr.

Reply:
@GTwhy Did you try this method?

https://docs.rs/minstant/latest/minstant/struct.Instant.html#method.checked_add

Reply:
@GTwhy Here is my result.

```
round:  1, minstant: 11.717 ns, stdinstant: 22.076 ns, systemtime: 21.868 ns
round:  2, minstant: 11.803 ns, stdinstant: 22.319 ns, systemtime: 22.058 ns
round:  3, minstant: 11.956 ns, stdinstant: 21.923 ns, systemtime: 22.226 ns
round:  4, minstant: 11.683 ns, stdinstant: 22.914 ns, systemtime: 22.186 ns
round:  5, minstant: 12.184 ns, stdinstant: 22.080 ns, systemtime: 22.227 ns
round:  6, minstant: 11.684 ns, stdinstant: 21.901 ns, systemtime: 21.965 ns
round:  7, minstant: 11.736 ns, stdinstant: 21.828 ns, systemtime: 21.791 ns
round:  8, minstant: 11.927 ns, stdinstant: 21.960 ns, systemtime: 22.308 ns
round:  9, minstant: 12.018 ns, stdinstant: 21.965 ns, systemtime: 21.972 ns
round: 10, minstant: 11.797 ns, stdinstant: 22.812 ns, systemtime: 22.392 ns
round: 11, minstant: 11.839 ns, stdinstant: 22.073 ns, systemtime: 21.913 ns
round: 12, minstant: 11.659 ns, stdinstant: 22.942 ns, systemtime: 22.182 ns
round: 13, minstant: 12.187 ns, stdinstant: 22.137 ns, systemtime: 22.176 ns
round: 14, minstant: 11.682 ns, stdinstant: 22.174 ns, systemtime: 22.480 ns
round: 15, minstant: 11.844 ns, stdinstant: 22.548 ns, systemtime: 22.182 ns
round: 16, minstant: 11.848 ns, stdinstant: 21.951 ns, systemtime: 22.226 ns
```


```rust
#![feature(bench_black_box)]

use std::hint::black_box;
use std::time::{Duration, SystemTime};

fn main() {
    let round = 16;
    for i in 1..=round {
        let timeout: Duration = Duration::new(1, 1);
        let t1 = minstant(timeout);
        let t2 = stdinstant(timeout);
        let t3 = systime(timeout);
        println!(
            "round: {i:>2}, \
            minstant: {t1:.3} ns, \
            stdinstant: {t2:.3} ns, \
            systemtime: {t3:.3} ns"
        );
    }
}

fn minstant(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = minstant::Instant::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn stdinstant(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = std::time::Instant::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn systime(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = SystemTime::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}
```

Reply:
We need to call `as_unix_nanos` after `now` or `checked_add` to make ddl consistent between machines.
```
fn minstant(timeout: Duration) -> f64 {
    let anchor = minstant::Anchor::new();
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = minstant::Instant::now().checked_add(timeout).unwrap().as_unix_nanos(&anchor);
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}
```
And my result:
```
round:  1, minstant: 55.486 ns, stdinstant: 32.512 ns, systemtime: 32.413 ns
round:  2, minstant: 55.215 ns, stdinstant: 32.367 ns, systemtime: 32.743 ns
round:  3, minstant: 55.253 ns, stdinstant: 32.512 ns, systemtime: 32.294 ns
round:  4, minstant: 55.538 ns, stdinstant: 32.378 ns, systemtime: 32.243 ns
round:  5, minstant: 55.021 ns, stdinstant: 32.388 ns, systemtime: 32.279 ns
round:  6, minstant: 56.239 ns, stdinstant: 32.661 ns, systemtime: 32.458 ns
round:  7, minstant: 56.171 ns, stdinstant: 32.601 ns, systemtime: 32.857 ns
round:  8, minstant: 55.251 ns, stdinstant: 32.387 ns, systemtime: 32.229 ns
round:  9, minstant: 55.682 ns, stdinstant: 32.689 ns, systemtime: 32.766 ns
round: 10, minstant: 55.858 ns, stdinstant: 33.068 ns, systemtime: 32.435 ns
round: 11, minstant: 55.517 ns, stdinstant: 32.535 ns, systemtime: 32.308 ns
round: 12, minstant: 55.317 ns, stdinstant: 32.999 ns, systemtime: 33.155 ns
round: 13, minstant: 55.481 ns, stdinstant: 32.567 ns, systemtime: 32.265 ns
round: 14, minstant: 55.401 ns, stdinstant: 32.449 ns, systemtime: 32.464 ns
round: 15, minstant: 56.879 ns, stdinstant: 33.406 ns, systemtime: 32.537 ns
round: 16, minstant: 55.543 ns, stdinstant: 32.707 ns, systemtime: 32.585 ns
```
It looks like your machine is faster than mine.
In my machine, minstant::Instant::now()@20ns; checked_add(timeout)@15ns; as_unix_nanos(&anchor)@15ns.
You can try to call `as_unix_nanos` after `checked_add` and check the time cost.

Reply:
Tests should be equivalent. See <https://docs.rs/serde/latest/src/serde/ser/impls.rs.html#611-625>.

@GTwhy Does your machine enable TSC? See <https://docs.rs/minstant/latest/minstant/fn.is_tsc_available.html>

```
round:  1, minstant: 12.146 ns, stdinstant: 33.616 ns, systime: 28.462 ns, minstant_unix_ts: 18.208 ns, systime_unix_ts: 43.620 ns, 
round:  2, minstant: 12.478 ns, stdinstant: 28.996 ns, systime: 28.499 ns, minstant_unix_ts: 18.206 ns, systime_unix_ts: 44.165 ns, 
round:  3, minstant: 12.039 ns, stdinstant: 29.259 ns, systime: 28.644 ns, minstant_unix_ts: 18.188 ns, systime_unix_ts: 43.722 ns, 
round:  4, minstant: 12.273 ns, stdinstant: 28.850 ns, systime: 28.459 ns, minstant_unix_ts: 18.188 ns, systime_unix_ts: 43.663 ns, 
round:  5, minstant: 12.035 ns, stdinstant: 28.703 ns, systime: 28.480 ns, minstant_unix_ts: 18.194 ns, systime_unix_ts: 43.633 ns, 
round:  6, minstant: 12.033 ns, stdinstant: 28.671 ns, systime: 28.456 ns, minstant_unix_ts: 18.173 ns, systime_unix_ts: 43.665 ns, 
round:  7, minstant: 12.137 ns, stdinstant: 28.799 ns, systime: 28.575 ns, minstant_unix_ts: 18.188 ns, systime_unix_ts: 46.622 ns, 
round:  8, minstant: 12.020 ns, stdinstant: 28.691 ns, systime: 28.519 ns, minstant_unix_ts: 18.190 ns, systime_unix_ts: 43.645 ns, 
round:  9, minstant: 12.077 ns, stdinstant: 28.737 ns, systime: 28.458 ns, minstant_unix_ts: 18.213 ns, systime_unix_ts: 43.729 ns, 
round: 10, minstant: 12.161 ns, stdinstant: 28.815 ns, systime: 28.716 ns, minstant_unix_ts: 18.213 ns, systime_unix_ts: 43.658 ns, 
round: 11, minstant: 12.142 ns, stdinstant: 28.825 ns, systime: 28.485 ns, minstant_unix_ts: 18.205 ns, systime_unix_ts: 43.542 ns, 
round: 12, minstant: 12.137 ns, stdinstant: 28.821 ns, systime: 28.471 ns, minstant_unix_ts: 18.193 ns, systime_unix_ts: 43.570 ns, 
round: 13, minstant: 12.130 ns, stdinstant: 28.875 ns, systime: 28.484 ns, minstant_unix_ts: 18.204 ns, systime_unix_ts: 43.526 ns, 
round: 14, minstant: 12.135 ns, stdinstant: 28.805 ns, systime: 28.467 ns, minstant_unix_ts: 18.192 ns, systime_unix_ts: 43.565 ns, 
round: 15, minstant: 12.914 ns, stdinstant: 28.819 ns, systime: 28.793 ns, minstant_unix_ts: 18.217 ns, systime_unix_ts: 43.586 ns, 
round: 16, minstant: 12.560 ns, stdinstant: 28.766 ns, systime: 28.729 ns, minstant_unix_ts: 18.214 ns, systime_unix_ts: 43.634 ns, 
```

```rust
#![feature(bench_black_box)]

use std::hint::black_box;
use std::time::{Duration, SystemTime};

fn main() {
    let round = 16;
    for i in 1..=round {
        let timeout: Duration = Duration::new(1, 1);
        let t1 = minstant(timeout);
        let t2 = stdinstant(timeout);
        let t3 = systime(timeout);
        let t4 = minstant_unix_ts(timeout);
        let t5 = systime_unix_ts(timeout);
        print!("round: {i:>2}, ");
        print!("minstant: {t1:.3} ns, ");
        print!("stdinstant: {t2:.3} ns, ");
        print!("systime: {t3:.3} ns, ");
        print!("minstant_unix_ts: {t4:.3} ns, ");
        print!("systime_unix_ts: {t5:.3} ns, ");
        println!();
    }
}

fn minstant(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = minstant::Instant::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn stdinstant(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = std::time::Instant::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn systime(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = SystemTime::now().checked_add(timeout).unwrap();
        let _ = black_box(ddl);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn minstant_unix_ts(timeout: Duration) -> f64 {
    let anchor = minstant::Anchor::new();
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = minstant::Instant::now().checked_add(timeout).unwrap();
        let ts: u64 = ddl.as_unix_nanos(&anchor);
        let _ = black_box(ts);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}

fn systime_unix_ts(timeout: Duration) -> f64 {
    let now = std::time::Instant::now();
    for _ in 0..1_000_000 {
        let timeout = black_box(timeout);
        let ddl = SystemTime::now().checked_add(timeout).unwrap();
        let ts: u64 = ddl
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap()
            .as_nanos() as u64;
        let _ = black_box(ts);
    }
    now.elapsed().as_nanos() as f64 / 1_000_000.0
}
```

Reply:
@Nugine Cool, that's my negligence.
TSC is enabled in my machine, but I'm not sure why it is so slow.
Is it because I used a VM ?

```
round:  1, minstant: 46.181 ns, stdinstant: 32.856 ns, systime: 32.830 ns, minstant_unix_ts: 53.505 ns, systime_unix_ts: 40.778 ns, 
round:  2, minstant: 45.700 ns, stdinstant: 32.724 ns, systime: 32.580 ns, minstant_unix_ts: 53.297 ns, systime_unix_ts: 41.198 ns, 
round:  3, minstant: 46.085 ns, stdinstant: 32.731 ns, systime: 32.651 ns, minstant_unix_ts: 53.546 ns, systime_unix_ts: 41.983 ns, 
round:  4, minstant: 45.895 ns, stdinstant: 32.822 ns, systime: 32.797 ns, minstant_unix_ts: 53.554 ns, systime_unix_ts: 40.946 ns, 
round:  5, minstant: 45.960 ns, stdinstant: 32.815 ns, systime: 32.698 ns, minstant_unix_ts: 53.815 ns, systime_unix_ts: 41.038 ns, 
round:  6, minstant: 45.737 ns, stdinstant: 32.794 ns, systime: 32.840 ns, minstant_unix_ts: 53.883 ns, systime_unix_ts: 41.730 ns, 
round:  7, minstant: 46.487 ns, stdinstant: 33.019 ns, systime: 32.694 ns, minstant_unix_ts: 53.466 ns, systime_unix_ts: 41.092 ns, 
round:  8, minstant: 45.712 ns, stdinstant: 32.748 ns, systime: 32.745 ns, minstant_unix_ts: 53.608 ns, systime_unix_ts: 40.963 ns, 
round:  9, minstant: 45.931 ns, stdinstant: 32.818 ns, systime: 32.784 ns, minstant_unix_ts: 54.226 ns, systime_unix_ts: 41.168 ns, 
round: 10, minstant: 46.221 ns, stdinstant: 32.986 ns, systime: 33.016 ns, minstant_unix_ts: 53.940 ns, systime_unix_ts: 41.254 ns, 
round: 11, minstant: 45.927 ns, stdinstant: 32.683 ns, systime: 32.797 ns, minstant_unix_ts: 53.594 ns, systime_unix_ts: 40.945 ns, 
round: 12, minstant: 45.736 ns, stdinstant: 32.679 ns, systime: 32.639 ns, minstant_unix_ts: 53.383 ns, systime_unix_ts: 40.933 ns, 
round: 13, minstant: 45.932 ns, stdinstant: 32.991 ns, systime: 32.723 ns, minstant_unix_ts: 53.304 ns, systime_unix_ts: 40.886 ns, 
round: 14, minstant: 45.668 ns, stdinstant: 32.684 ns, systime: 33.002 ns, minstant_unix_ts: 53.945 ns, systime_unix_ts: 41.302 ns, 
round: 15, minstant: 46.088 ns, stdinstant: 32.992 ns, systime: 32.806 ns, minstant_unix_ts: 53.445 ns, systime_unix_ts: 40.687 ns, 
round: 16, minstant: 45.950 ns, stdinstant: 32.757 ns, systime: 32.650 ns, minstant_unix_ts: 53.284 ns, systime_unix_ts: 40.777 ns,
```

Reply:
> Is it because I used a VM ?

Maybe. I ran the test on a physical machine.